### DISCUSSION TITLE
proj_ll2ra.csh produces different ranges of values.

### DISCUSSION THREAD

**Discussion by RenzoAnc** [gmtsar/gmtsar#487] [Q&A]
Hello everyone. I have a question about proj_ll2ra.csh script.

I'm using this script to transform *.grd files (in lat/lon coordinates) which were obtained after some steps I applied to correct the unwrap.grd files (obtained with unwrap_intf.csh, SBAS technique) through GACOS.

After GACOS correction, I could obtained all *.grd files with the same lat/lon as you can see in the following images.

**Case 1 (2016304_2016328):**

![Case1](https://user-images.githubusercontent.com/105986961/195909185-b569e2d5-0586-4ca4-a0dc-cb68703529ef.png)

**Case 2 (2016304_2016352):**

![Case2](https://user-images.githubusercontent.com/105986961/195909208-e2403a2c-69db-4fb9-8c25-5b685a116c1f.png)

but I couldn't obtain the same values for ran/azi coordinates range when I applied proj_ll2ra.csh. 

**Case 1 (2016304_2016328):**

![Case1a](https://user-images.githubusercontent.com/105986961/195909534-6fb7458c-5b24-4819-b668-091c90531bfe.png)

**Case 2 (2016304_2016352):**

![Case2a](https://user-images.githubusercontent.com/105986961/195909547-f5dcc996-d232-4b07-bfc4-fd1749daac04.png)

I don't know what is the problem. For proj_ll2ra.csh script, I used the trans.dat file created in topo directory (when intf_tops.csh one.in batch_tops.config command is executed). I know it will generate problems when I execute sbas command, so I'll be thankful to you if you give me an idea about it.

Thanks in advance. 

Regards.

**Comment by Xiaohua-Eric-Xu** [comment]
The final range is determined by this line here
```
60 gmt xyz2grd rap `gmt gmtinfo rap -I32/64 -bi3f` -I32/64 -r -G$3 -bi3f
```
where `rap` is range/azimuth/phase that comes from `gmt grdtracking` the projection grid with 
```
 52 gmt grdtrack llp -nl -Gllr.grd -bi3f -bo4f > llpr
 53 gmt grdtrack llpr -nl -Glla.grd -bi4f -bo5f > llpra
```
where `lla.grd and llr.grd` are projection grids computed from trans.dat which should be the same for every pair based on what you showed. An `llp` is basically the table converted from your input grid. 

Do note the final range is determined over the valid part of `rap` by `gmt gmtinfo rap -I32/64 -bi3f`, so if `rap` certain records are NaN or missing somewhere, the range will change.

**Reply by RenzoAnc** [reply]
Hello, professor @Xiaohua-Eric-Xu. Thank you for your reply. I checked the script out and I've realized the dimensions of pixel size of unw_F.grd file (unwrap.grd corrected with GACOS, in lat/lon coordinates) I obtained, which is used in **proj_ll2ra.csh trans.dat unw_F.grd unw.grd**, is 0.000831951348258721168 in horizontal and 0.0008318479144385111657 in vertical direction. They're different of 0.00083333333333 value, which is mentioned in proj_ll2ra.csh script. Could it be affecting the results with proj_ll2ra.csh?. 

![Lines](https://user-images.githubusercontent.com/105986961/195997587-7ad5c601-f9ae-4027-a61b-510bba0a70b9.png)

Regards.

**Reply by Xiaohua-Eric-Xu** [reply]
I don't think they will. My suspicion is that there being some missing values in the input grid. Is there a way that you could share the two input grid and your trans.dat? e.g. Google Drive?

**Reply by RenzoAnc** [reply]
Sure. I add the link of Google Drive below, where you can find the *.grd  (unw_F.grd) files and trans.dat file. The *.grd files are inside of different directories (2016304_2016328 and 2016304_2016352).

Link: https://drive.google.com/drive/folders/127sUF27m4-5VjyocnRXXH_7vvOAkUHi7?usp=sharing

Regards, professor @Xiaohua-Eric-Xu.

**Reply by Xiaohua-Eric-Xu** [reply]
I think the two grids has different coverage. Look at the left side at the far range, the right one has values at further range thus after projection it has a larger max range.

![image](https://user-images.githubusercontent.com/16844111/196503321-b3a66244-d98e-4b85-a043-4eb0a6ff7c7a.png)

**Reply by RenzoAnc** [reply]
I get you, professor @Xiaohua-Eric-Xu. I've realized these fringes appear in left and bottom side when I applied **proj_ra2ll.csh trans.dat unwrap.grd unwrap_ll.grd**, where unwrap.grd file was obtained with intf_tops.csh. I applied it because I need to have the unwrap.grd files in geographic coordinates for GACOS correction. After GACOS correction and some steps more, I get the unw_F.grd files I sent you before.

I show you the unwrap.grd files visualized in *.pdf files below (both were obtained with intf_tops.csh).

Directory 2016304_2016328:


![unwrap_304328](https://user-images.githubusercontent.com/105986961/196544787-2c61a6a5-9c6d-478a-8d5c-c483cdff77b6.png)

Directory 2016304_2016352:


![unwrap_304352](https://user-images.githubusercontent.com/105986961/196544805-078e6e09-45fe-4287-ab5a-613aaee7a876.png)


I don't know if these fringes appears before or after applying the **proj_ra2ll.csh script**. In unwrap.pdf files, you can see continue colors in right and top sides, but I think it is a GMT feature. Possibly I'm wrong.

After I apply **proj_ra2ll.csh** program, these fringes appear. You can see it below (in QGIS program).


Directory 2016304_2016328:
![unwrap_ll_304328](https://user-images.githubusercontent.com/105986961/196544706-142f3bdf-5b1f-4f45-9a93-71949e2d4f48.png)


Directory 2016304_2016352:

![unwrap_ll_304352](https://user-images.githubusercontent.com/105986961/196544717-6ce686d5-b761-4fd9-87fa-4d2bade79ca7.png)

These are the results which appear in terminal after applying **proj_ra2ll.csh** (It's the same for both *.grd files).

File:
[results_script_304328.txt](https://github.com/gmtsar/gmtsar/files/9815127/results_script_304328.txt)

In the file, you can see a line: "ls: No match". Maybe it's the problem.

Thank you for advance.

Regards.

**Reply by Xiaohua-Eric-Xu** [reply]
I guess one way of solving this inconsistency is to cut your unwrapped phase to an area with all valid inputs and then run the script for atm correction?

**Reply by RenzoAnc** [reply]
I'll process these results taking in account your advice, professor . If I have another question or news, I'll inform you. Regards.

**Reply by RenzoAnc** [reply]
Hello, professor @Xiaohua-Eric-Xu. Sorry for the late response. About the unwrap.grd problem, it was resolved. I cut the unwrap.grd file and it works now (the results have the same range values). But this problem still be in the corr.grd file. I have a question, ¿which configure parameters in the batch_tops.config do the files (*.grd) resolution change which are generate with intf_tops.csh?. I'm looking for the right parameters which generate an unwrap_ll.grd file with a 0.000833333 resolution approximately (after applying unwrap_intf.csh and proj_ra2ll.csh to results of intf_tops.csh).

Regards.

**Reply by RenzoAnc** [reply]
Hi, professor. About the problem with corr.grd file, you can reproduce it using the files in the link of Google Drive below: corr_F.grd, unwrap_F.grd and trans.dat. 

https://drive.google.com/drive/folders/1bc8lLHf0rPRv0auL5tE-tK_Dz0bCs9Y4?usp=sharing

In QGIS (or other software), you can prove corr_F.grd and unwrap_F.grd cover the same area and they have the same resolution, but when you applied proj_ll2ra.csh trans.dat unwrap_F.grd (or corr_F.grd) unwrap.grd, the range values are different. 

Unwrap.grd file:

![I2](https://user-images.githubusercontent.com/105986961/200053628-9f58a2a6-cc8e-403c-8c27-314222e7376d.jpg)

Corr.grd file:

![I1](https://user-images.githubusercontent.com/105986961/200053641-80dcd567-4b4a-42db-a10f-278c8f4ed6ff.jpg)

The problem was resolved for unwrap.grd files when I generated files that cover the same area. This idea didn't work for corr.grd file.

Thank you in advance.

Regards.

**Reply by Xiaohua-Eric-Xu** [reply]
How did you produce the corr.grd? If it is from InSAR processing using GMTSAR, the range should have no negative values.

**Reply by RenzoAnc** [reply]
Hello, professor. I obtained corr.grd (final) following the same steps I did to get unwrap.grd (final). The difference is the GACOS correction step, which I didn't apply to corr.grd (initial) file. After some days working on it, I could get files limited with the same range values.

You can see an example of unwrap_G.grd and corr_G.grd files obtained after GACOS correction.

unwrap_G.grd file:

![unwrap](https://user-images.githubusercontent.com/105986961/201425634-28f40ed5-8f0e-4bb7-b358-7cf5522c638a.jpg)



corr_G.grd file:

![corr](https://user-images.githubusercontent.com/105986961/201425638-fb472485-b718-460c-a0de-beffb6e5b190.jpg)


Unfortunately, the results with sbas command are wrong. I show you my results after applying sbas command and the results in the terminal.

vel.png file:

![vel_ll](https://user-images.githubusercontent.com/105986961/201424834-fd9c0508-036e-44f9-916a-8e16787dcf17.jpg)

Ubuntu terminal:
 
![Terminal](https://user-images.githubusercontent.com/105986961/201425773-3e47eae8-dcb0-4a4a-82f3-d9c38d404e0f.png)

I don't know what is the problem. Do I need to change another file to run sbas command?, besides the corr.grd and unwrap.grd files.

To run sbas command, I just worked with the unwrap_G.grd and corr_G.grd files obtained after applying GACOS correction. In the processing (after applying intf_tops.csh and unwrap_intf.csh), I just replaced unwrap.grd and corr.grd with unwrap_G.grd and corr_G.grd files.

Regards, professor.

**Reply by RenzoAnc** [reply]
After applying GACOS correction, the range values and other parameters changed.


![Change](https://user-images.githubusercontent.com/105986961/201426775-90be7b66-f9bd-43ee-b33a-fa1322d2ac8f.jpg)

**Reply by Xiaohua-Eric-Xu** [reply]
I don't think it's a problem related to sbas. The input dimension (79 8) looks a bit suspicious to me. How come that they are so small. Did you downsample your unwrapped interferogram by a lot? Why did you do that?

**Comment by RenzoAnc** [comment]
Hello, professor. Yes, the input dimensions of corr.grd and unwrap.grd files were altered during the processing.

In the unwrap.grd file case, I did the following steps:

1) I converted the unwrap.grd file to a unwrap_r.grd file using gmt grdcut.
   I did this step to remove the continuous fringes you can see in the comments before. Using this step, I could obtained same range values for unwrap.grd and corr.grd files.
   It changed the initial dimensions:  375 x 450 -> 307 x 230.
   
   
2) I converted the unwrap_r.grd (radar) file to unwrap_ll.grd (lat/lon) file using proj_ra2ll.csh script. I did this step because I need to get a *.grd file in lat/lon coordinates to process with GACOS correction. In this step, I used the gauss_350 file which was created after executing int_tops.csh command. This gauss file is necessary to get the unwrap_ll.grd file with a resolution of 0.000833333 (for GACOS).
   It changed the dimensions too: 307 x 230 -> 110 x 90.
   
   
3) I did some additional steps using others softwares to process with GACOS. These steps were done to get the unwrap file in *.tiff format, because I did the GACOS correction using the script main.m which uses it. 


![GACOS](https://user-images.githubusercontent.com/105986961/201479647-1e81021a-2173-414d-9d6e-aac2515911d9.jpg)





At the end of this step, I could obtained a unw_F.grd file corrected with GACOS.
   In this case, the dimensions weren't altered.
   
   
4) Finally, I got a unwrap_G file (radar) of the unw_F.grd file (lat/lon) obtained in the step before. For do it, I used the proj_ll2ra.csh script.
   It changed the dimensions: 110 x 90 -> 79 x 8 .
   
   
In the corr.grd file case, I did the same steps without considering the third step. It's logical because the corr.grd file doesn't need to be corrected with GACOS.

I've realized the proj_ra2ll.csh and proj_ll2ra.csh altered so much the dimensions of *.grd files. Is there a way to get a file with similar dimension values after applying proj_ll2ra.csh script (fourth step)?

Regards

**Reply by Xiaohua-Eric-Xu** [reply]
Yes, indeed this is very small range (307x230). Do you have a plot of your unwrap.grd? Also when you run proj_ra2ll.csh, you could select a projection resolution. And for proj_ll2ra.csh you'll have to change the `-I` option inside the script to increase the resolution.

**Reply by RenzoAnc** [reply]
Hello, professor. I've taken into account your advices and I did the changes I mention below:


**a) To apply gmt grdcut:**

I've delimited other region (in -R option) to get a *.grd output file bigger.

Dimensions: 375x450 -> 332x405




![Comparing](https://user-images.githubusercontent.com/105986961/202241778-66586b9f-e864-40c4-b426-62eef3313f45.jpg)



**b) In the proj_ra2ll.csh  script:**

b.1) In the m2s.csh script:

Initial command:

set dx = `gmt math -Q $dx 10 MUL  = `
set dy = `gmt math -Q $dy 10 MUL  = `

Final command:

set dx = `gmt math -Q $dx 100 MUL  = `
set dy = `gmt math -Q $dy 100 MUL  = `

They define the values of incs[2] which are used in proj_ra2ll.csh script, in the -I option.

Dimensions: 332x405 -> 200x300


**c) In the proj_ll2ra.csh script:**


Initial command:

gmt xyz2grd rap `gmt gmtinfo rap -I32/64 -bi3f` -I32/64 -r -G$3 -bi3f

Final command:

gmt xyz2grd rap `gmt gmtinfo rap -I32/64 -bi3f` -I8/4 -r -G$3 -bi3f

Dimensions: 200x300 -> 340x224

As you can see, I've changed the -I option values to configure the dimensions of output files (*.grd). Are the changes correct?. Which values of final dimensions shall you recommend me to get a good result after applying sbas command?. In this case, I was able to get a *.grd with this dimensions: 340x224.

In the last case (proj_ll2ra.csh), you can see I obtained a better resolution what I had initially (200x300 -> 340x224).

Regards.

**Reply by RenzoAnc** [reply]
I show you a result of the corr.grd file with the dimensions 340 x224. It looks strange.
![corr](https://user-images.githubusercontent.com/105986961/202276407-856efa7d-72ff-4c2c-af81-2248ff472611.jpg)

This problem is generated when I execute pro_ll2ra.csh script. The *.grd output is good when I don't change any line in the script. But If I change the values 32/64 in the -I option, the output looks like the image above.

**Reply by Xiaohua-Eric-Xu** [reply]
The spaces among pixels is because you are projecting from lower resolution to higher resolution. All projection methods require some level of re-mosaicing which will reduce resolution (even just by a tiny bit). So when you go from lower resolution to higher resolution, it'll create null mosaics.

**Reply by RenzoAnc** [reply]
Thank you for your reply, professor. I have another question. It's about **sbas command**. If I apply this command, which pixels will I get after applying it?. I've applied this command with unwrap.grd files which had many pixels (low coherence > 0.08), but after sbas command, many pixels dissapears in the disp_*.grd and vel.grd files. I think the coherence is affecting the results. Is there a way to change this coherence threshold to get more pixels after applying sbas command?. Regards.

Before applying sbas command (an unwrap.grd file):
![07Dec2022](https://user-images.githubusercontent.com/105986961/206281672-561d7d62-451b-41e6-8efb-aa301f1d5d5f.jpg)


After applying sbas command (disp_*.grd file):
![07Dec2022_2](https://user-images.githubusercontent.com/105986961/206281685-da91ba76-55b7-4db0-9dad-d19159945f4d.jpg)

**Reply by Xiaohua-Eric-Xu** [reply]
SBAS will remove any pixel that has any decorrelation (i.e. NaN in any unwrap.grd). In one unwrap.grd, your decorrelation may look like the plot above, but what if you superimpose all of your unwrap.grd? It'll probably look like the second plot. The main reason is because decorrelation happens inconsistently through space and time. You may wanna use a common stacked-coherence mask, and set the unwrapping threshold very small when running snaphu[_interp].csh.

**Reply by RenzoAnc** [reply]
Hello, professor. I've changed the unwrapping threshold (0.08 -> 0.03) and I got more pixels in the result after sbas. I think it's a small value to analyse the displacements and velocity. For example, SNAP software uses 0.7 as a coherence threshold to analyze the displacement results. In GMTSAR, which are the most common values during the processing?. 

![vel_ll](https://user-images.githubusercontent.com/105986961/206486141-1d301792-a805-4ba5-ac80-077682184137.jpg)


About using a common stacked-coherence mask, Should be it used before snaphu.csh?. I mean if I should put a common corr.grd file in all <date1,date2> directories (after moving or eliminating the remaining corr.grd files) which contains the pixels I want to get in the result, and I apply snaphu.csh after it.

Or only use a common corr.grd, for sbas processing, in the second column of intf.tab file.

**Reply by Xiaohua-Eric-Xu** [reply]
> For example, SNAP software uses 0.7 as a coherence threshold to analyze the displacement results. In GMTSAR, which are the most common values during the processing?.

I think SNAP uses a different definition of coherence. In GMTSAR we typically use 0.1 unless one creates a common coherence mask (then one could use as small as possible). The latter should be the case for time-series analysis with coherence loss like yours.
